### PR TITLE
Make generic pipeline work with airflow2x

### LINKED ISSUE #3166: make generic pipelines to DAG rendering work with Airflow 2.x
**Describe the issue**
Currently, even when using generic pipeline editor with python files and notebooks, the rendered DAG is not compatible with Airflow 2..6.0 or higher, or any Airflow 2.x, for that matter :-)
`
Broken DAG: [/opt/airflow/dags/repo/SvensWetterapprootpvc-0202162036.py] Traceback (most recent call last): File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed File "/opt/airflow/dags/repo/SvensWetterapprootpvc-0202162036.py", line 2, in <module> from airflow.contrib.kubernetes.volume_mount import VolumeMount ModuleNotFoundError: No module named 'airflow.contrib.kubernetes'`

**To Reproduce**
1. Create a generic pipeline
2. Submit it to the Airflow runtime with Airflow 2.x

**Screenshots or log output**
If applicable, add screenshots or log output to help explain your problem.
<details><summary>Log Output</summary>
<pre>
Paste the log output here.
</pre>
</details>

**Expected behavior**
DAg is rendered correctly for use in Airflow 2.x

Starting with jinja template
https://github.com/elyra-ai/elyra/blob/main/elyra/templates/airflow/airflow_template.jinja2

and elyra/elyra/pipeline/airflow/processor_airflow.py

possible hints
https://airflow.apache.org/docs/apache-airflow-providers-cncf-kubernetes/stable/operators.html

https://airflow.apache.org/docs/apache-airflow/2.5.3/howto/upgrading-from-1-10/index.html

https://airflow.apache.org/docs/apache-airflow/2.5.3/core-concepts/dags.html

https://github.com/apache/airflow/pull/21653

**Deployment information**
Describe what you've deployed and how:
 - Elyra version: latest
 - Installation source: [e.g. PyPI, conda, from source, official container image, custom container image]
 - Deployment type: Open Data Hub pre-RedHat takeover
 - Operating system: [e.g. macos, linux]

### PR DESCRIPTION
**Pull request by shalberd**
fixes #3166 

@ianonavy @lresende NOT intended to work with Airflow 1.x

### What changes were proposed in this pull request?

changes to airflow processor and airflow pipeline template with regards to Airflow 2.8.2 or higher support
also added pendulum instead of days_ago since deprecated

https://github.com/apache/airflow/pull/21653/files

Conceptual overlap with a fork, came to the same code in parallel

change to the Kubernetes client SDK in generic pipeline part of template since the Airflow abstractions
were all deprecated and removed except for Secret.

Finally, Airflow 2 adds logic that makes config_file mutually exclusive
with in_cluster, so we need to ensure that None is passed as None and
not string "None".

See also

https://github.com/kflow-ai/elyra/commit/f9d132954e008d30145f18794aa543d97f121a5f#diff-dc6c3f666aad9271fa5e9b8c31e3f0582cd39a7d2516cbc2240731fe4456e641

### How was this pull request tested?
In contrast to kubeflow pipelines, even for Airflow 1.x and the different pipeline editors, there do not seem to be any tests.
I'd like to test the built wheel file in a docker image in conjunction with ODH.
Mostly seeing whether the generated DAG code works with Airflow 2.8.2 and higher.
 
Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

### CODE REVIEW

**Comment by ianonavy** [comment]
One thing I forgot to note is that `from airflow.providers.cncf.kubernetes.operators.kubernetes_pod import KubernetesPodOperator` was deprecated in favor of `from airflow.providers.cncf.kubernetes.operators.pod import KubernetesPodOperator` in Airflow 2.6.0. Decided to keep the older one for compatibility with earlier releases; not sure which version we are targeting.

**Comment by shalberd** [comment]
Regarding KubernetesPodOperator location: Yes, [good point.](https://github.com/apache/airflow/blob/v2-6-stable/airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py#L25) Well, 2.5.x is already well past, plus there were some security bugs in 2.5.x that someone in our org mentioned that make it a good idea to go further than 2.5.x. Airflow 2.6.0 came out in May, so we can assume it is current.

**Comment by giladd123** [comment]
I did approximatly the same thing in an air tight environment so it would be a problem for me to commit my code but I think I can help.
I also encountered a problem with the resources, what seemed to work for me is switching `resources = ... ` to something like:
```
{% if operation.cpu_request or operation.mem_request or operation.gpu_limit %}
container_resources=k8s.V1ResourceRequirements(
    requests={
        {% if operation.cpu_request %}
        'cpu': '{{ operation.cpu_request }}',
        {% endif %}
        {% if operation.mem_request %}
        'memory': '{{ operation.mem_request }}G',
        {% endif %}
    },
    limits={
        {% if operation.cpu_request %}
        'cpu': '{{ operation.cpu_request }}',
        {% endif %}
        {% if operation.mem_request %}
        'memory': '{{ operation.mem_request }}G',
        {% endif %}
        {% if operation.gpu_limit %}
        'nvidia.com/gpu': '{{ operation.gpu_limit }}', # or 'amd.com/gpu'
        {% endif %}
    },
)
{% endif %}
```
In this case, the UI doesn't have an option for CPU and Memory limits, so the limits could probably be removed (In my k8s cluster a limit must be set so I just set it in some ratio of the request).

I lack the knowledge to change the UI to include limits but I'm willing to help with anything else!

**Comment by shalberd** [comment]
@giladd123 aggreed, setting limits is good practice, even when not enforced by [LimitRange](https://docs.openshift.com/container-platform/4.10/nodes/clusters/nodes-cluster-limit-ranges.html#nodes-cluster-limit-ranges-limits_nodes-cluster-limit-ranges) and max ratio request to limit.

[Quoting](https://www.opensourcerers.org/2022/06/20/the-importance-of-getting-resource-requests-and-limits-right/)

```
Requests and Limits should never be treated as min/max requirements for any pod. The requested amount should be a reflection of the pod’s required resources (both CPU and memory) under load. The ratio between request and limit should be as close to 1.0 as possible. Any divergence from this should have a valid business reason. The request amount reflects the pod’s normal usage, the limit is there as a safety net to protect the rest of the cluster’s load, in case it misbehaves, or a mistake was made during the profiling of the application.
```
For the template, setting limits to either equal to requests or minimally higher by a factor of x 1.2 would be good practice and make the cluster as a whole more stable. In the end, developers sets those requests in the GUI and they must be admonished to set realistic request sizes. With a limit set, they'll realize soon they are wrong in their pipeline steps definition for resources.

**Comment by giladd123** [comment]
> For the template, setting limits to either equal to requests or minimally higher by a factor of x 1.2 would be good practice and make the cluster as a whole more stable.

Altough being a good practice, I don't think setting a hard ratio is a good idea as in the end this gives the user less flexability (for example, my cluster enforces a 1:4 ratio on cpu request and limit, with a 1.2x jobs will just not go up). I don't think we should force good practices on users, we should let them deal with this themselves.

**Comment by shalberd** [comment]
@giladd123 Agreed, however, not setting limits at all leads to pods being scheduled on a node that they really should not be scheduled on. In your example of 1 to 4 ratio, should all pods every run on the same node at the same time, you're in for huge trouble, having a node outage and re-scheduling, leading to interruptions.

As you mentioned, it'd be best to at least have the option to get limits at the GUI level. 
Then, each enterprise can make their developers input educated settings. 
I need to check whether cpu request, memory request, gpu request, gpu memory request are GUI required values. 
If they are, then the same (mandatory fields) should apply for limits on the GUI as well, being of course user-defineable.
Not providing limits is simply sloppy, I've got no idea why noone ever criticized that.

**Comment by lresende** [comment]
Where we are with this? I am assuming still a WIP and we are waiting for more changes before a full review?

**Comment by dolevi101** [comment]
Hi, is there any update?

**Comment by shalberd** [comment]
yes, it is WIP still, in large part due to the fact that I have not received any feedback, neither from ODH nor here, on why CPU and Memory limits are not in the variables and thus cannot be used at least optionally.

**Comment by shalberd** [comment]
I am leaving out the issue of resource limits for now. My aim is to test build this and integrate it into an open data hub notebook container, then test it with our Airflow 2.6.2 instance that is reading the Elyra-generated DAGs from our Gitlab company-internally.

**Comment by shalberd** [comment]
getting there. Muuch better. Just tested this an forgot about what Gilad told me back then regarding changes to resources assembly, 
Bildschirmfoto 2023-11-24 um 17 11 29
Bildschirmfoto 2023-11-24 um 17 11 53

https://airflow.apache.org/docs/apache-airflow-providers-cncf-kubernetes/stable/_modules/airflow/providers/cncf/kubernetes/operators/pod.html

making the changes now and will test the locally-built wheel file once more with Airflow 2.6.x then, next week.

i.e. in the airflow template file, for now only with that gpu limit, no cpu and memory limits as we do not have them yet in GUI.
```
        {% if operation.cpu_request or operation.mem_request or operation.gpu_limit %}
                                                            container_resources=k8s.V1ResourceRequirements(
                                                                requests={
                                                                    {% if operation.cpu_request %}
                                                                        'cpu': '{{ operation.cpu_request }}',
                                                                    {% endif %}
                                                                    {% if operation.mem_request %}
                                                                        'memory': '{{ operation.mem_request }}G',
                                                                    {% endif %}
                                                                },
                                                                limits={
                                                                    {% if operation.gpu_limit %}
                                                                        'nvidia.com/gpu': '{{ operation.gpu_limit }}', # or 'amd.com/gpu'
                                                                    {% endif %}
                                                                },
                                                            )
        {% endif %}
```
I don't use GPUs, but I want to find out before I make the whl again whether there is also a property for the kind of gpu, don't wanna hard-code either `nvidia.com/gpu` or `amd.com/gpu`
Another thing I noticed: CPU always needs to be an integer greater than 1, memory also. This is not good. We should later fix that GUI and property part to more K8S and Openshift style resource units: https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/

File: etc/generic/requirements-elyra.txt
Comment (shalberd) [inline etc/generic/requirements-elyra.txt:20]: this is probably wrong: pendulum is not required for Elyra itself, but when running the rendered DAG definition on Airflow. And Airflow as it seems already has the library in its worker and scheduler containers. Or at least my runtime containers have it. Need to research a little about this.

Comment (shalberd) [reply etc/generic/requirements-elyra.txt]: @romeokienzler ready for review

**Comment by shalberd** [comment]
squashed commits as well to be more readable

**Comment by lresende** [comment]
It would be good to have a 2nd person to validate this is working before we look into merging it...

**Comment by kevin-bates** [comment]
@shalberd - thank you for this pull request.  I am surprised at how little changes there are and don't see any kind of Airflow version checking happening.  With these changes, is elyra agnostic to the Airflow version or do these changes _replace_ the existing Airflow 1x support with Airflow 2x support?  I guess I was expecting to see a new subclass (e.g., `AirflowPipelineProcessor2`) or runtime type (`APACHE_AIRFLOW_2`) so as to preserve existing support.  If a replacement - what do we tell our existing Airflow 1x users?

**Comment by shalberd** [comment]
@kevin-bates it is a replacement for Airflow 2.x. I have already talked with @lresende in a Friday community call and we have the idea so far to make this part of Elyra 4.x, i.e. no more Airflow 1.x support soon. About lifecycle management: I only have contact with Astronomer and the Airflow community helm chart maintainer @thesuperzapper ... my judgment call would be: Airflow 1.x is long deprecated, no more security updates ... announce Airflow 1.x support derelease with Elyra 4.x

**Comment by MR-GOYAL** [comment]
@shalberd I have tried changes as mentioned in the [fork](https://github.com/shalberd/elyra/tree/make_generic_pipeline_work_with_airflow2x) and its working fine with the airflow 2.7.2(version we are using). I have tested in our environment and airflow pipeline are executing as expected.
Thanks for the airflow 2.x support on elyra

**Comment by romeokienzler** [comment]
waiting for #3202 to be merged

**Comment by lresende** [comment]
Looks great that the generic pipelines are working with these changes, we need to make sure custom components made for Airflow 2.x are also working well.

File: elyra/templates/airflow/airflow_template.jinja2
Comment (shalberd) [inline elyra/templates/airflow/airflow_template.jinja2:25]: @lresende @MR-GOYAL @kevin-bates @ianonavy interesting ... from Airflow 2.7.0 on, the secrets line changed to

The one notable exception is that we will **continue to support the airflow.providers.cncf.kubernetes.secret.Secret class**.

at https://airflow.apache.org/docs/apache-airflow/2.7.0/howto/upgrading-from-1-10/index.html#step-5-upgrade-airflow-dags

vs. before 2.7.0

The one notable exception is that **we will continue to support the airflow.kubernetes.secret.Secret class**.

at https://airflow.apache.org/docs/apache-airflow/2.6.2/howto/upgrading-from-1-10/index.html#step-5-upgrade-airflow-dags

Now, with Airflow 2.8.2, `from airflow.kubernetes.secret import Secret` is still working fine, but I should probably change it soon to `from airflow.providers.cncf.kubernetes.secret import Secret`

**Comment by shalberd** [comment]
confirming this here is still working beautifully for Airflow 2.8.2

**Comment by shalberd** [comment]
@lresende the readthedocs latest: I searched for all mentions of Airflow there and made some changes to the documentation, pertaining to Airflow 2 support vs. Airflow 1 support, the scope of Airflow 2 support in Elyra 4.

**Comment by shalberd** [comment]
Hi @lresende I had to fix some lint-server E226 missing whitespace around arithmetic operator code in elyra/tests, but all well now.

**Comment by shalberd** [comment]
@lresende getting an error at step validate runtime images 

"=> ERROR: Container image pytorch/pytorch:2.5.1-cuda11.8-cudnn9-devel is not a suitable Elyra runtime image
make[1]: *** [Makefile:326: validate-runtime-image] Error 1"

**Comment by shalberd** [comment]
@caponetto looking much better, your change worked for the validate images step. Nice job, thanks again.

You mentioned that cypress component, do you know why this check fails in the step "run integration tests":

Screenshot 2024-12-20 at 20 57 19

**Comment by caponetto** [comment]
@shalberd From my observations over the past few months, I believe there is a flaky integration test. Sometimes it passes, sometimes it fails. You can try rerunning the job a few times to see if it passes.

**Comment by shalberd** [comment]
> From my observations over the past few months, I believe there is a flaky integration test. Sometimes it passes, sometimes it fails. You can try rerunning the job a few times to see if it passes.

Yes, I amended my commit with nothing and force-pushed again, leading to the tests to run through again (I do not have enough rights to re-run checks from the Github GUI).

All tests, including that flaky yarn cypress integration test, ran through now. Wonder if we should have a separate issue filed for that flaky yarn cy integration test ....

**Comment by shalberd** [comment]
About other people running this Airflow 2 specific code to generate DAG python code:

Another person from India, thanks, @MR-GOYAL, also verified earlier this year with an earlier Airflow version that it worked fine. https://github.com/elyra-ai/elyra/pull/3167#issuecomment-1873860868

**Comment by caponetto** [comment]
> Wonder if we should have a separate issue filed for that flaky yarn cy integration test

I believe so. Also, if the failure is too frequent to the point of slowing work down, I suggest disabling it until its stability is restored.

**Comment by shalberd** [comment]
Hi @lresende I rebased my feature branch with the latest changes to main, i.e. your refactor of [pipeline implementation and namings](https://github.com/elyra-ai/elyra/pull/3160) It should not have any impact on my code changes here.

File: docs/source/user_guide/pipeline-components.md
Comment (lresende) [inline docs/source/user_guide/pipeline-components.md:31]: @shalberd is this a conflict?

Comment (shalberd) [reply docs/source/user_guide/pipeline-components.md:31]: yes, it looks like a rebase tool merge conflict, resolving it now. Thank you, resolving it now.

**Review by lresende — APPROVED** [review]

**Pull request merged**
